randovania/resolver/game_state.py

```diff
@@ -0,0 +1,212 @@
+    def apply_damage(self, damage: int) -> Self:
+        return self
+
+    @override
+    def apply_node_heal(self, resources: ResourceCollection) -> Self:
+        return self
+
+    @override
+    def debug_string(self, resources: ResourceCollection) -> str:
+        return ""
+
+    @override
+    def limited_by_maximum(self, resources: ResourceCollection) -> Self:
+        return self
+
+    @override
+    def apply_collected_resource_difference(
+        self, new_resources: ResourceCollection, old_resources: ResourceCollection
+    ) -> Self:
+        return self
+
+    @override
+    def apply_new_starting_resource_difference(
+        self, new_resources: ResourceCollection, old_resources: ResourceCollection
+    ) -> Self:
+        return self
+
+    @override
+    def resource_requirements_for_satisfying_damage(self, damage: int) -> list[ResourceRequirement]:
+        return []
+
+
+class EnergyTankGameState(GameState):
+    _energy: int
+    _starting_energy: int
+    _energy_per_tank: int
+    _energy_tank: ItemResourceInfo
+    _resource_database: ResourceDatabase
+    _region_list: RegionList
+
+    def __init__(
+        self, starting_energy: int, energy_per_tank: int, resource_database: ResourceDatabase, region_list: RegionList
+    ):
+        self._energy = starting_energy
+        self._starting_energy = starting_energy
+        self._energy_per_tank = energy_per_tank
+        self._energy_tank = resource_database.energy_tank
+        self._resource_database = resource_database
+        self._region_list = region_list
+
+    @override
+    def resource_database(self) -> ResourceDatabase:
+        return self._resource_database
+
+    @override
+    def region_list(self) -> RegionList:
+        return self._region_list
+
+    def _maximum_energy(self, resources: ResourceCollection) -> int:
+        num_tanks = resources[self._energy_tank]
+        return self._starting_energy + (self._energy_per_tank * num_tanks)
+
+    @override
+    def health_for_damage_requirements(self) -> int:
+        return self._energy
+
+    @override
+    def is_better_than(self, other: GameState | None) -> bool:
+        if other is None:
+            return True
+        assert isinstance(other, EnergyTankGameState)
+        return self._energy > other._energy
+
+    @override
+    def apply_damage(self, damage: int) -> Self:
+        result = copy.copy(self)
+        result._energy -= damage
+        return result
+
+    @override
+    def apply_node_heal(self, resources: ResourceCollection) -> Self:
+        return self.limited_by_maximum(resources)
+
+    @override
+    def debug_string(self, resources: ResourceCollection) -> str:
+        return f"{self._energy}/{self._maximum_energy(resources)} Energy"
+
+    @override
+    def limited_by_maximum(self, resources: ResourceCollection) -> Self:
+        result = copy.copy(self)
+        result._energy = min(result._energy, result._maximum_energy(resources))
+        return result
```

Comment:
Is this supposed to use `max` or just return  `_maximum_energy` instead of doing this?

Reply:
I think that kinda depends what this is supposed to do?
With the way its used, it sounds like it should just be `result._energy = result._maximum_energy(resources)` with the function being renamed to "get_maximum" or something

But with the docstring saying: `Applies any resource limit` it sounds more like this is called with an over-maximum-energy which needs to be capped back.

Reply:
Lots of the uses ended expecting it to "fully heal", which is actually a behaviour I wanted to leave game-specific, but then didn't actually implement it. But yes, I designed this to be just "please cap health at the maximum health".